curriculum/challenges/english/08-coding-interview-prep/rosetta-code/k-d-tree.md

```diff
@@ -0,0 +1,354 @@
+---
+id: 5a23c84252665b21eecc7ecb
+title: K-d tree
+challengeType: 5
+---
+
+## Description
+<section id='description'>
+  A k-d tree (short for <i>k</i>-dimensional tree) is a space-partitioning data structure for organizing points in a k-dimensional space. 
+  k-d trees are a useful data structure for several applications, such as searches involving a multidimensional search key (e.g. range searches and nearest neighbor searches). 
+  k-d trees are a special case of binary space partitioning trees. k-d trees are not suitable, however, for efficiently finding the nearest neighbor in high dimensional spaces. As a general rule, if the dimensionality is <i>k</i>, the number of points in the data, <i>N</i>, should be <i>N</i> ≫ 2<sup><i>k</i></sup>.
+  Otherwise, when k-d trees are used with high-dimensional data, most of the points in the tree will be evaluated and the efficiency is no better than exhaustive search, and other methods such as approximate nearest-neighbor are used instead.
+  Write a function to perform a nearest neighbour search using k-d tree. The function takes two parameters: an array of k-dimensional points, and a single k-dimensional point whose nearest neighbour should be returned by the function. A k-dimensional point will be given as an array of k elements.
```

Comment:
One new thing we're trying to do is move the instructions to the instructions section below, usually the last sentence or two in each of the challenge descriptions that often begin with "Write a function...":

```
## Instructions
<section id='instructions'>
Write a function to perform a nearest neighbour search using k-d tree. The function takes two parameters: an array of k-dimensional points, and a single k-dimensional point whose nearest neighbour should be returned by the function. A k-dimensional point will be given as an array of k elements.
 </section>
```

Sometimes it's not possible since if the description is really short or everything reads like instructions, so in those cases we can just leave all the text in the description section.